docs/extensibility.md

```diff
@@ -1,364 +1,148 @@
+    async fn discover(
+        &self,
+        request: tonic::Request<DiscoverRequest>,
+    ) -> Result<Response<Self::DiscoverStream>, Status> {
+        // Get the discovery url from the `DiscoverRequest` else return an error signifying 
+        // an improper HTTP Configuration
+        let url = match request.get_ref().discovery_details.get("discoveryEndpoint") {
+            Some(url) => url.clone(),
+            None => { return Err(tonic::Status::new(
+                tonic::Code::InvalidArgument,
+                format!("Invalid HTTP discovery handler ... discovery_details [{:?}] do not contain a url at key 'discoveryEndpoint'", request.get_ref().discovery_details),
+            )); }
+        };
+        // Create a channel for sending and receiving device updates
+        let (mut stream_sender, stream_receiver) = mpsc::channel(4);
+        let mut register_sender = self.register_sender.clone();
+        tokio::spawn(async move {
+            loop {
+                let resp = get(&url).await.unwrap(); 
+                // Response is a newline separated list of devices (host:port) or empty
+                let device_list = &resp.text().await.unwrap();
+                let devices = device_list
+                    .lines()
+                    .map(|endpoint| {
+                        let mut properties = HashMap::new();
+                        properties.insert(BROKER_NAME.to_string(), "http".to_string());
+                        properties.insert(DEVICE_ENDPOINT.to_string(), endpoint.to_string());
+                        Device {
+                            id: endpoint.to_string(),
+                            properties,
+                            mounts: Vec::default(),
+                            device_specs: Vec::default(),
+                        }
+                    })
+                    .collect::<Vec<Device>>();
+                // Send the Agent the list of devices.
+                if let Err(_) = stream_sender.send(Ok(DiscoverResponse { devices })).await {
+                    // Agent dropped its end of the stream. Stop discovering and signal to try to re-register.
+                    register_sender.send(()).await.unwrap();
+                    break;
+                }
+            }
+        });
+        // Send the agent one end of the channel to receive device updates
+        Ok(Response::new(stream_receiver))
+    }
 ```
-
-Now you are ready to **build the HTTP broker**!  To do so, we simply need to run this step from the base folder of the Akri repo:
+### Build the DiscoveryHandler container
+Now you are ready to build your HTTP discovery handler and push it to your container registry. To do so, we simply need to run this step from the base folder of the Akri repo:
 
 ```bash
 HOST="ghcr.io"
 USER=[[GITHUB-USER]]
-BROKER="http-broker"
+DH="http-discovery-handler"
 TAGS="v1"
 
-IMAGE="${HOST}/${USER}/${BROKER}:${TAGS}"
+DH_IMAGE="${HOST}/${USER}/${DH}:${TAGS}"
 
 docker build \
```

Comment:
I haven't been able to get the [Dockerfile](https://github.com/kate-goldenring/akri-discovery-handler-template/blob/main/Dockerfile.discovery-handler) to build. I'd love to have second eyes on why I'm getting a "'./target/release/deps/http-discovery-handler*': No such file or directory" error.

Reply:
did you ever build the source code?  i got that error, but hadn't actually built the new discovery handler (so there was no target folder).

Reply:
also, if you change `IMAGE` to `DH_IMAGE`, you should update `docker build --tag=${IMAGE}`

Reply:
i don't think that the target/release/deps will have an entry for the thing being built ... only the dependencies (this is purely observational though ... i have no real understanding).  previously for this sort of thing, we created a dummy main.rs for the first build and deleted ./target/release/.fingerprint/akri-http-discovery-handler* before the second build.

Reply:
here's a good blog: https://alexbrand.dev/post/how-to-package-rust-applications-into-minimal-docker-containers/

Reply:
nevermind most of my comments.  i think the problem is that you are using COPY (uppercase) to move a file from one place in the container to another (typically COPY is used to move a file into a container from another container or the host)...i tried this and it worked:

```
FROM amd64/rust:1.47 as build
RUN rustup component add rustfmt --toolchain 1.47.0-x86_64-unknown-linux-gnu
RUN USER=root cargo new --bin dh
WORKDIR /dh
COPY ./Cargo.toml ./Cargo.toml
RUN cargo build --release
RUN rm ./src/*.rs
RUN rm ./target/release/deps/akri_http_discovery_handler*
COPY ./src ./src
RUN cargo build --release
RUN cp ./target/release/akri-http-discovery-handler /akri-http-discovery-handler
ENTRYPOINT ["/akri-http-discovery-handler"]
```

Reply:
you may want to a smaller run container:
```
FROM amd64/rust:1.47 as build
RUN rustup component add rustfmt --toolchain 1.47.0-x86_64-unknown-linux-gnu
RUN USER=root cargo new --bin dh
WORKDIR /dh
COPY ./Cargo.toml ./Cargo.toml
RUN cargo build --release
RUN rm ./src/*.rs
RUN rm ./target/release/deps/akri_http_discovery_handler*
COPY ./src ./src
RUN cargo build --release

FROM amd64/debian:buster-slim
COPY --from=build /dh/target/release/akri-http-discovery-handler /akri-http-discovery-handler
ENTRYPOINT ["/akri-http-discovery-handler"]
```

Reply:
Thank you! This got me running and pointed out that {{project-name}} needed to be adjusted to swap `-` and `_` [when removing the dependencies](https://github.com/kate-goldenring/akri-discovery-handler-template/blob/40d302230a297df342b5c05f76500df8313b8257/Dockerfile.discovery-handler#L8).